library/core/src/fmt/mod.rs

```diff
@@ -971,9 +971,14 @@ pub trait UpperHex {
 /// `p` formatting.
 ///
 /// The `Pointer` trait should format its output as a memory location. This is commonly presented
-/// as hexadecimal.
+/// as hexadecimal. For more information on formatters, see [the module-level documentation][module].
 ///
-/// For more information on formatters, see [the module-level documentation][module].
+/// Printing of pointers is not a reliable way to discover how Rust programs are implemented,
+/// because the act of reading an address of a value may change how the value is stored in memory,
+/// and prevent some code optimizations.
+///
+/// Rust may move values in memory, and reuse the same memory locations for different values, so
+/// the printed pointer values are not guaranteed to be stable nor unique identifiers of objects.
```

Comment:
We may be able to handwave slightly less by being more direct: The problem is more fundamentally that _printing the pointer is part of the program_. Therefore by inserting such prints to obtain introspection on their program's behavior, a programmer now introspects a different program. There's no _guarantee_ of equivalence in machine code emitted for two similar programs.